### PR TITLE
[WIP] Pull streams

### PR DESCRIPTION
**Pull request by dignifiedquire**
Migrate 

- [x] `fixed-size-chunker`
- [x] `exporter`
- [x] `importer`

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -2,7 +2,7 @@ sudo: false
 language: node_js
 node_js:
   - 4
-  - 5
+  - stable
```
Comment (diasdavid) [inline .travis.yml]: Add Node.js 5 tests too

File: src/exporter.js [hunk 2]
```diff
@@ -1,124 +1,45 @@
...
-log.err = debug('unixfs:error')
 const isIPFS = require('is-ipfs')
-const UnixFS = require('ipfs-unixfs')
-const series = require('run-series')
-const Readable = require('readable-stream').Readable
-const pathj = require('path')
-const util = require('util')
-const fieldtrip = require('field-trip')
-const cleanMultihash = require('./clean-multihash')
+const traverse = require('pull-traverse')
```
Comment (noffle) [inline src/exporter.js]: Slick, I didn't know about `pull-traverse`!

File: src/exporter.js [hunk 3]
```diff
@@ -1,124 +1,45 @@
...
-    return new Exporter(hash, dagService, options)
-  }
+const dirExporter = require('./exporters/dir')
+const fileExporter = require('./exporters/file')
 
+module.exports = (hash, dagService, options) => {
   // Sanitize hash
   if (!isIPFS.multihash(hash)) {
     throw new Error('not valid multihash')
   }
```
Comment (diasdavid) [inline src/exporter.js]: This should be part of `cleanMultihash`

**Comment by diasdavid** [comment]
Got an error, basically due to assumed ordering

```

  1) core importer small file (smaller than a chunk) inside a dir:

      Uncaught AssertionError: expected [ Array(3) ] to deeply equal [ Array(3) ]
      + expected - actual

           "path": "foo/bar/200Bytes.txt"
           "size": 211
         }
         {
      +    "multihash": "Qmf5BQbTUyUAvd6Ewct83GYGnE1F6btiC3acLhR8MDxgkD"
      +    "path": "foo/bar"
      +    "size": 270
      +  }
      +  {
           "multihash": "QmQrb6KKWGo8w7zKfx2JksptY6wN7B2ysSBdKZr4xMU36d"
           "path": "foo"
           "size": 320
         }
      -  {
      -    "multihash": "Qmf5BQbTUyUAvd6Ewct83GYGnE1F6btiC3acLhR8MDxgkD"
      -    "path": "foo/bar"
      -    "size": 270
      -  }
       ]

```

**Comment by dignifiedquire** [comment]
@diasdavid what do we want to do about the ordering, just make the tests not care about it?

**Comment by diasdavid** [comment]
Also phantom fails on the last test of Importer with

`PhantomJS 2.1.1 (Linux 0.0.0) WARN: 'Unhandled rejection: AssertionError: expected [ Array(5) ] to deeply equal [ Array(5) ]`

**Comment by dignifiedquire** [comment]
Why would this be a bug? it's just traversing/emitting in a different order

**Comment by diasdavid** [comment]
In PhantomJS

```
 1) IPFS data importing tests on the Browser importer nested directory (2 levels deep):
     timeout of 80000ms exceeded. Ensure the done() callback is being called in this test.

 
  2) IPFS data importing tests on the Browser chunker: fixed size 256 KiB chunks of non scalar filesize:
     expected [ Array(4) ] to have a length of 5 but got 4
```

**Comment by diasdavid** [comment]
Also first time seeing this:

```
[BABEL] Note: The code generator has deoptimised the styling of "/Users/david/code/js-ipfs-unixfs-engine/test/test-fixed-size-chunker.js" as it exceeds the max of "100KB".
[BABEL] Note: The code generator has deoptimised the styling of "/Users/david/code/js-ipfs-unixfs-engine/test/test-exporter.js" as it exceeds the max of "100KB".
[BABEL] Note: The code generator has deoptimised the styling of "/Users/david/code/js-ipfs-unixfs-engine/test/test-importer.js" as it exceeds the max of "100KB".

```

**Comment by dignifiedquire** [comment]
I have no idea what phantomjs has for a problem, except that it might be too slow to do anything right :( I can't debug Phantom on my machine currently as every time I start it, it will die with 

```
phantomjs(11121,0x7fff793ce000) malloc: *** error for object 0x7fc0898d2400: pointer being freed was not allocated
*** set a breakpoint in malloc_error_break to debug
```

**Comment by dignifiedquire** [comment]
@diasdavid fixed the first issue

**Comment by dignifiedquire** [comment]
found the reason of the second issue: https://www.npmjs.com/package/buffer#one-minor-difference and pull-block relies on this :( not sure how this worked before as the chunker in `master` uses buffer.slice as well.

**Comment by diasdavid** [comment]
PhantomJS exists in the pipeline to verify that the ES5 transpiled version works as expected, in this specific case it is not a transpilation issue, but an 'old browser' problem. We haven't declared what are the official minimum browser version to run js-ipfs, but it will be certainly a browser that at least requires WebRTC (which brings it to the class of modern browsers, where we won't hit this issue).

I've been wanting to be able to disable PhantomJS for a while (https://github.com/dignifiedquire/aegir/issues/48), seems that now it is the time :)

**Pull request merged**